### PR TITLE
Json API for Ord

### PR DESCRIPTION
**Pull request by ynohtna92**
Adds several json API endpoints following the API sketch @casey derived in https://github.com/casey/ord/pull/1662.

See [docs/src/guides/restapi.md](https://github.com/ynohtna92/ord-litecoin/blob/json-api/docs/src/guides/restapi.md) for full list of changes.

Things still to do:
- Fixes tests so they pass! Failing due to the wallet client having to be injected into server.rs resulting in mismatch of chain type when running tests
- Figure out how errors will be handled and shown. They need to be wrapped in json but only when accept_json.0 is true.

### CODE REVIEW

**Comment by lnky79** [comment]
Nice job you have done, thanks a lot.

File: src/subcommand/server.rs [hunk 1]
```diff
@@ -426,16 +485,33 @@ impl Server {
 
     let inscriptions = index.get_inscriptions_on_output(outpoint)?;
 
-    Ok(
+    Ok(if accept_json.0 {
+      axum::Json(serde_json::json!({
+        "value": output.value,
```
Comment (jun0tpyrc) [inline src/subcommand/server.rs:490]: for json there is no inscriptions info like webpage && ( Sat Range info etc)


how about 
```
+        "inscriptions": inscriptions.iter().map(|inscription| {
+          serde_json::json!({
+            "href": format!("/inscription/{}", inscription),
+          })
+        }).collect::<Vec<_>>(),
```
etc here ?


e.g.
https://ordinals.com/output/4ee303f05452a63c2578e42df1b3c1b163a98746182b8f07a8d33706287659d3:0
^ have inscriptions & other infos

```
{"_links":{"self":{"href":"/output/4ee303f05452a63c2578e42df1b3c1b163a98746182b8f07a8d33706287659d3:0"},"transaction":{"href":"/tx/4ee303f05452a63c2578e42df1b3c1b163a98746182b8f07a8d33706287659d3"}},"address":"bc1q3xfux78s6hr2g5h3rgpqwlmwqqfczt3r20k644","script_pubkey":"OP_0 OP_PUSHBYTES_20 8993c378f0d5c6a452f11a02077f6e0013812e23","transaction":"4ee303f05452a63c2578e42df1b3c1b163a98746182b8f07a8d33706287659d3","value":546}
```
not yet

File: src/subcommand/server.rs [hunk 2]
```diff
@@ -503,31 +580,110 @@ impl Server {
...
     let inscription = index.get_inscription_by_id(txid.into())?;
 
     let blockhash = index.get_transaction_blockhash(txid)?;
 
-    Ok(
+    let transaction = index
+      .get_transaction(txid)?
+      .ok_or_not_found(|| format!("transaction {txid}"))?;
+
+    Ok(if accept_json.0 {
```
Comment (jun0tpyrc) [inline src/subcommand/server.rs:622]: Seem no input/output infos?
https://ordinals.com/tx/4ee303f05452a63c2578e42df1b3c1b163a98746182b8f07a8d33706287659d3

```
> curl "XXXXXXX/ord/tx/4ee303f05452a63c2578e42df1b3c1b163a98746182b8f07a8d33706287659d3?apikey=pFuDmQca4vwhroAC8vc7o8k73PkhBUkwgaHYMq4wa2MpxA4fiF"  -H "accept: application/json" -s | jq
{
  "_links": {
    "block": null,
    "inputs": [
      {
        "href": "/output/b6d8221b942281fa4ea03fd7364b1c3ed2b6b82ba27c7726d7a0ccaec532e94a:23"
      }
    ],
    "inscription": {
      "href": "/inscription/4ee303f05452a63c2578e42df1b3c1b163a98746182b8f07a8d33706287659d3i0"
    },
    "outputs": [
      {
        "href": "/output/4ee303f05452a63c2578e42df1b3c1b163a98746182b8f07a8d33706287659d3:0"
      }
    ],
    "self": {
      "href": "/tx/4ee303f05452a63c2578e42df1b3c1b163a98746182b8f07a8d33706287659d3"
    }
  },
  "transaction": "4ee303f05452a63c2578e42df1b3c1b163a98746182b8f07a8d33706287659d3"
}
```

File: src/subcommand/server.rs (hunk 1 above)
Comment (jun0tpyrc) [reply src/subcommand/server.rs:490]: after this change  = 
```
{
  "_links": {
    "self": {
      "href": "/output/4ee303f05452a63c2578e42df1b3c1b163a98746182b8f07a8d33706287659d3:0"
    },
    "transaction": {
      "href": "/tx/4ee303f05452a63c2578e42df1b3c1b163a98746182b8f07a8d33706287659d3"
    }
  },
  "address": "bc1q3xfux78s6hr2g5h3rgpqwlmwqqfczt3r20k644",
  "inscriptions": [
    {
      "href": "/inscription/4ee303f05452a63c2578e42df1b3c1b163a98746182b8f07a8d33706287659d3i0"
    }
  ],
  "script_pubkey": "OP_0 OP_PUSHBYTES_20 8993c378f0d5c6a452f11a02077f6e0013812e23",
  "transaction": "4ee303f05452a63c2578e42df1b3c1b163a98746182b8f07a8d33706287659d3",
  "value": 546
}
```

Comment (ynohtna92) [reply src/subcommand/server.rs:490]: Will add this!

**Comment by nibty** [comment]
Looking good. It looks like this will require the wallet RPC to be enabled. Is it possible to make that optional? I run a read-only bitcoind and ord. Also, are there plans to add pagination to the feed endpoint? Thanks.

**Comment by ynohtna92** [comment]
> Looking good. It looks like this will require the wallet RPC to be enabled. Is it possible to make that optional? I run a read-only bitcoind and ord. Also, are there plans to add pagination to the feed endpoint? Thanks.

I can try to make the wallet rpc optional, with a flag in the ord.yaml.

No plans for pagination in the feed endpoint at this stage. I do my pagination by using the inscriptions/ endpoint and changed it to output 1000 at a time.

**Comment by nibty** [comment]
> No plans for pagination in the feed endpoint at this stage. I do my pagination by using the inscriptions/ endpoint and changed it to output 1000 at a time.

I didn't see the inscription endpoint. That works for pagination.

BTW. Some inscriptions cause 502s.

> example
```
curl -H "Accept: application/json" http://XXXXX.XXX/inscription/994731268d65fa8698dbba9580d7980f7c04d79633d8287baa4de264b915488ei0
<html><body><h1>502 Bad Gateway</h1>
The server returned an invalid or incomplete response.
</body></html>
```

```
thread 'tokio-runtime-worker' panicked at 'called `Result::unwrap()` on an `Err` value: UnrecognizedScript', src/subcommand/server.rs:1048:81
stack backtrace:
rust_begin_unwind
at /rustc/84c898d65adf2f39a5a98507f1fe0ce10a2b8dbc/library/std/src/panicking.rs:579:5
core::panicking::panic_fmt
at /rustc/84c898d65adf2f39a5a98507f1fe0ce10a2b8dbc/library/core/src/panicking.rs:64:14
core::result::unwrap_failed
at /rustc/84c898d65adf2f39a5a98507f1fe0ce10a2b8dbc/library/core/src/result.rs:1750:5
core::result::Result<T,E>::unwrap
<F as axum::handler::Handler<(M,T1,T2,T3,T4),S,B>>::call::{{closure}}
```

**Comment by ynohtna92** [comment]
> 994731268d65fa8698dbba9580d7980f7c04d79633d8287baa4de264b915488ei0

I think it is because it should have an address but it doesn't for some reason....

File: src/subcommand/server.rs (hunk 1 above)
Comment (ynohtna92) [reply src/subcommand/server.rs:490]: https://github.com/casey/ord/pull/2036/commits/3f44a1afe35135cd52c60200917d2d85a8b54f9b

**Comment by ynohtna92** [comment]
All tests should now pass.

Added a api_wallet_enable flag to ord.yaml so that the wallet endpoints can be used.

Looking for code reviewers and testers @raphjaph

Not a rust dev normally so there could be some things that are not best practice.

Ie I'm not able to simplify this code snippet. Can't seem to do `client.as_ref().unwrap()` as it performs a move on Client so had to put it inside of a if let instead.

`if let Some(client) = client.as_ref() { ... }`

**Review by bigcrypSv — APPROVED** [review]

**Comment by nibty** [comment]
Found another 502 on this one 994731268d65fa8698dbba9580d7980f7c04d79633d8287baa4de264b915488ei0

Jun 15 16:48:53 xxx ord[276030]: thread 'tokio-runtime-worker' panicked at 'called `Result::unwrap()` on an `Err` value: UnrecognizedScript', src/subcommand/server.rs:1086:81
Jun 15 16:48:53 xxx ord[276030]: stack backtrace:
Jun 15 16:48:53 xxx ord[276030]:    0: rust_begin_unwind
Jun 15 16:48:53 xxx ord[276030]:              at /rustc/84c898d65adf2f39a5a98507f1fe0ce10a2b8dbc/library/std/src/panicking.rs:579:5
Jun 15 16:48:53 xxx ord[276030]:    1: core::panicking::panic_fmt
Jun 15 16:48:53xxx ord[276030]:              at /rustc/84c898d65adf2f39a5a98507f1fe0ce10a2b8dbc/library/core/src/panicking.rs:64:14
Jun 15 16:48:53 xxx ord[276030]:    2: core::result::unwrap_failed
Jun 15 16:48:53 xxx ord[276030]:              at /rustc/84c898d65adf2f39a5a98507f1fe0ce10a2b8dbc/library/core/src/result.rs:1750:5
Jun 15 16:48:53 xxx ord[276030]:    3: ord::subcommand::server::Server::inscription::{{closure}}

**Comment by ynohtna92** [comment]
> This is a great start! I want to start looking at this but this PR is a bit too large to review. I would also like to start simple with only the endpoints that do not require an API key or use the `/wallet` endpoint. Additionally I think having an address to inscription numbers table will make the database too large. If you feel like cutting out the stuff that would be great, otherwise I will try to do some slicing and dicing myself on the weekend.

Yeah I got carried away with adding things to this pr. The address stuff can be removed or alternatively enabled with a flag similar to index-sats. The address table is quite small compared to the sat ranges table if you are concerned with database size, I think more importantly it could be a performance hit depending on how well it searches and drains though the number arrays when adding or removing an inscription from an address. The wallet API stuff could also be it in its own pr as well. Not sure how to turn this into 3 PRS since two parts depend on the first.

**Comment by lnky79** [comment]
> This is a great start! I want to start looking at this but this PR is a bit too large to review. I would also like to start simple with only the endpoints that do not require an API key or use the `/wallet` endpoint. Additionally I think having an address to inscription numbers table will make the database too large. If you feel like cutting out the stuff that would be great, otherwise I will try to do some slicing and dicing myself on the weekend.

Great to hear your kind response @raphjaph on the API issue. 

Actually, I am working on some UTXO/PSBT signer library for Ordinals marketplaces which needs **UTXO/cardinal selecting** by seller/buyer address before generating unsigned PSBT for users. And I have used the branch created by @ynohtna92 for 1 month, it works pretty well on positive number inscriptions. 

In my view, the `/address` endpoint or its official index table is very important for the **transfer protection** for user inscription UTXOs in the ecosystem. Hope it can be preserved finally, just as @ynohtna92 said, we can have some option like `--index-address` to make things happen.

Thanks for the awesome job you guys have done.

**Comment by Mathieu-Be** [comment]
Hey guys, is anyone still working on this ? Very interested by this so would be glad to contribute and make a smaller PR for non auth endpoints

**Comment by raphjaph** [comment]
We've started work on the JSON API, feel free to add more endpoints similiar to how we have designed the existing ones :)

**Pull request closed without merging**